text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,205 @@
+- Feature Name: guaranteed_slice_repr
+- Start Date: 2025-02-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC guarantees the in-memory representation of slice and str references.
+Specifically, `&[T]` and `&mut [T]` are guaranteed to have the same layout as:
+
+```rust
+#[repr(C)]
```

Comment:
For the calling convention slices are actually passed as two separate arguments rather than as a single struct argument. Depending on the calling convention the latter can cause it to be passed on the stack, which would likely be a perf regression.

Reply:
It’s clarified later in the guide level explanation that the RFC does not intend to say anything about argument passing, only memory layout:

> The precise ABI of slice references is not guaranteed, so `&[T]` may not be
passed by-value or returned by-value from an `extern "C" fn`.

Reply:
I think the RFC is right to only propose anything for memory layout and nothing for calling conventions. There's much more room for regret when committing to a particular calling convention for slices than when committing to a memory layout. And there's also much less upside: if you're passing a slice by value over FFI, it's a *local* decision to take it apart and pass the pointer and length however you want. In contrast, if the memory layout of a slice somewhere in memory isn't what you need, you may have to change a `&[T]` field in a `repr(C)` struct to a pointer-length pair. That can affect a much larger amount of code far removed from the FFI calls.

Reply:
@Jules-Bertholet I think we would rather simply not promise anything right now about the calling convention of these types, certainly not that it gets split across arguments, because such
1. changes the arity between C and Rust, which is quite painful to reason about
2. closes the door on a future C extension introducing proper slice types and us being ABI compatible with them